File: csharp/Hron/Schedule.cs
using Hron.Ast;
using Hron.Cron;
using Hron.Eval;
using HronParser = Hron.Parser.Parser;
using HronDisplay = Hron.Display.Display;

namespace Hron;

/// <summary>
/// The main entry point for parsing and evaluating hron schedule expressions.
/// </summary>
/// <example>
/// <code>
/// var schedule = Schedule.Parse("every weekday at 9:00 except dec 25 in America/New_York");
/// var next = schedule.NextFrom(DateTimeOffset.Now);
/// if (next.HasValue)
/// {
///     Console.WriteLine($"Next occurrence: {next.Value}");
/// }
/// </code>
/// </example>
public sealed class Schedule
{
    private readonly ScheduleData _data;
    private readonly TimeZoneInfo _zoneInfo;

    private Schedule(ScheduleData data, TimeZoneInfo zoneInfo)
    {
        _data = data;
        _zoneInfo = zoneInfo;
    }

    /// <summary>
    /// Parses an hron expression into a Schedule.
    /// </summary>
    /// <param name="input">The hron expression</param>
    /// <returns>The parsed schedule</returns>
    /// <exception cref="HronException">If the input is invalid</exception>
    public static Schedule Parse(string input)
    {
        var data = HronParser.Parse(input);
        var zoneInfo = ResolveTimezone(data.Timezone);
        return new Schedule(data, zoneInfo);
    }

    /// <summary>
    /// Converts a 5-field cron expression to a Schedule.
    /// </summary>
    /// <param name="cronExpr">The cron expression</param>
    /// <returns>The parsed schedule</returns>
    /// <exception cref="HronException">If the cron expression is invalid</exception>
    public static Schedule FromCron(string cronExpr)
    {
        var data = CronConverter.FromCron(cronExpr);
        var zoneInfo = ResolveTimezone(data.Timezone);
        return new Schedule(data, zoneInfo);
    }

    /// <summary>
    /// Validates an hron expression without throwing.
    /// </summary>
    /// <param name="input">The hron expression</param>
    /// <returns>True if the expression is valid</returns>
    public static bool Validate(string input)
    {
        try
        {
            HronParser.Parse(input);
            return true;
        }
        catch (HronException)
        {
            return false;
        }
    }

    /// <summary>
    /// Computes the next occurrence after the given time.
    /// </summary>
    /// <param name="now">The reference time</param>
    /// <returns>The next occurrence, or null if none exists</returns>
    public DateTimeOffset? NextFrom(DateTimeOffset now)
    {
        // Convert now to the schedule's timezone
        var nowInTz = TimeZoneInfo.ConvertTime(now, _zoneInfo);
        return Evaluator.NextFrom(_data, nowInTz, _zoneInfo);
    }

    /// <summary>
    /// Computes the next n occurrences after the given time.
    /// </summary>
    /// <param name="now">The reference time</param>
    /// <param name="count">The number of occurrences to compute</param>
    /// <returns>A list of the next n occurrences</returns>
    public IReadOnlyList<DateTimeOffset> NextNFrom(DateTimeOffset now, int count)
    {
        var nowInTz = TimeZoneInfo.ConvertTime(now, _zoneInfo);
        return Evaluator.NextNFrom(_data, nowInTz, count, _zoneInfo);
    }

    /// <summary>
    /// Computes the most recent occurrence strictly before the given time.
    /// </summary>
    /// <param name="now">The reference time (exclusive upper bound)</param>
    /// <returns>The previous occurrence, or null if none exists</returns>
    public DateTimeOffset? PreviousFrom(DateTimeOffset now)
    {
        var nowInTz = TimeZoneInfo.ConvertTime(now, _zoneInfo);
        return Evaluator.PreviousFrom(_data, nowInTz, _zoneInfo);
    }

    /// <summary>
    /// Checks if a datetime matches this schedule.
    /// </summary>
    /// <param name="dateTime">The datetime to check</param>
    /// <returns>True if the datetime matches</returns>
    public bool Matches(DateTimeOffset dateTime)
    {
        var dtInTz = TimeZoneInfo.ConvertTime(dateTime, _zoneInfo);
        return Evaluator.Matches(_data, dtInTz, _zoneInfo);
    }

    /// <summary>
    /// Returns a lazy enumerable of occurrences starting after the given time.
    /// </summary>
    /// <param name="from">The reference time (exclusive)</param>
    /// <returns>An enumerable of occurrences</returns>
    public IEnumerable<DateTimeOffset> Occurrences(DateTimeOffset from)
    {
        var fromInTz = TimeZoneInfo.ConvertTime(from, _zoneInfo);
        return Evaluator.Occurrences(_data, fromInTz, _zoneInfo);
    }

    /// <summary>
    /// Returns a lazy enumerable of occurrences where from &lt; occurrence &lt;= to.
    /// </summary>
    /// <param name="from">The start time (exclusive)</param>
    /// <param name="to">The end time (inclusive)</param>
    /// <returns>An enumerable of occurrences in the range</returns>
    public IEnumerable<DateTimeOffset> Between(DateTimeOffset from, DateTimeOffset to)
    {
        var fromInTz = TimeZoneInfo.ConvertTime(from, _zoneInfo);
        var toInTz = TimeZoneInfo.ConvertTime(to, _zoneInfo);
        return Evaluator.Between(_data, fromInTz, toInTz, _zoneInfo);
    }

    /// <summary>
    /// Converts this schedule to a 5-field cron expression.
    /// </summary>
    /// <returns>The cron expression</returns>
    /// <exception cref="HronException">If the schedule cannot be expressed as cron</exception>
    public string ToCron() => CronConverter.ToCron(_data);

    /// <summary>
    /// Returns the IANA timezone name, or null if not specified.
    /// </summary>
    public string? Timezone => string.IsNullOrEmpty(_data.Timezone) ? null : _data.Timezone;

    /// <summary>
    /// Returns the canonical string representation of this schedule.
    /// </summary>
    public override string ToString() => HronDisplay.Render(_data);

    /// <summary>
    /// Returns the underlying schedule data.
    /// </summary>
    public ScheduleData Data => _data;

    /// <summary>
    /// Resolve timezone, defaulting to UTC for deterministic behavior.
    /// </summary>
    private static TimeZoneInfo ResolveTimezone(string? tzName)
    {
        if (string.IsNullOrEmpty(tzName))
        {
            return TimeZoneInfo.Utc;
        }

        // Try IANA timezone ID first
        try
        {
            return TimeZoneInfo.FindSystemTimeZoneById(tzName);
        }
        catch (TimeZoneNotFoundException)
        {
            // On Windows, may need to try TZConvert or alternative ID
            // .NET 6+ should support IANA IDs on all platforms
            throw HronException.Parse($"unknown timezone: {tzName}", new Span(0, 0), tzName);
        }
    }
}

File: java/src/main/java/io/hron/Schedule.java
package io.hron;

import io.hron.ast.ScheduleData;
import io.hron.cron.CronConverter;
import io.hron.display.Display;
import io.hron.eval.Evaluator;
import io.hron.parser.Parser;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * The main entry point for parsing and evaluating hron schedule expressions.
 *
 * <p>Example usage:
 *
 * <pre>{@code
 * Schedule schedule = Schedule.parse("every weekday at 9:00 except dec 25 in America/New_York");
 * Optional<ZonedDateTime> next = schedule.nextFrom(ZonedDateTime.now());
 * if (next.isPresent()) {
 *     System.out.println("Next occurrence: " + next.get());
 * }
 * }</pre>
 */
public final class Schedule {
  private final ScheduleData data;
  private final ZoneId zoneId;

  private Schedule(ScheduleData data, ZoneId zoneId) {
    this.data = data;
    this.zoneId = zoneId;
  }

  /**
   * Parses an hron expression into a Schedule.
   *
   * @param input the hron expression
   * @return the parsed schedule
   * @throws HronException if the input is invalid
   */
  public static Schedule parse(String input) throws HronException {
    ScheduleData data = Parser.parse(input);
    ZoneId zoneId = resolveTimezone(data.timezone());
    return new Schedule(data, zoneId);
  }

  /**
   * Converts a 5-field cron expression to a Schedule.
   *
   * @param cronExpr the cron expression
   * @return the parsed schedule
   * @throws HronException if the cron expression is invalid
   */
  public static Schedule fromCron(String cronExpr) throws HronException {
    ScheduleData data = CronConverter.fromCron(cronExpr);
    ZoneId zoneId = resolveTimezone(data.timezone());
    return new Schedule(data, zoneId);
  }

  /**
   * Validates an hron expression without throwing.
   *
   * @param input the hron expression
   * @return true if the expression is valid
   */
  public static boolean validate(String input) {
    try {
      Parser.parse(input);
      return true;
    } catch (HronException e) {
      return false;
    }
  }

  /**
   * Computes the next occurrence after the given time.
   *
   * @param now the reference time
   * @return the next occurrence, or empty if none exists
   */
  public Optional<ZonedDateTime> nextFrom(ZonedDateTime now) {
    // Convert now to the schedule's timezone
    ZonedDateTime nowInTz = now.withZoneSameInstant(zoneId);
    return Evaluator.nextFrom(data, nowInTz, zoneId);
  }

  /**
   * Computes the next n occurrences after the given time.
   *
   * @param now the reference time
   * @param n the number of occurrences to compute
   * @return a list of the next n occurrences
   */
  public List<ZonedDateTime> nextNFrom(ZonedDateTime now, int n) {
    ZonedDateTime nowInTz = now.withZoneSameInstant(zoneId);
    return Evaluator.nextNFrom(data, nowInTz, n, zoneId);
  }

  /**
   * Computes the most recent occurrence strictly before the given time.
   *
   * @param now the reference time (exclusive upper bound)
   * @return the previous occurrence, or empty if none exists
   */
  public Optional<ZonedDateTime> previousFrom(ZonedDateTime now) {
    ZonedDateTime nowInTz = now.withZoneSameInstant(zoneId);
    return Evaluator.previousFrom(data, nowInTz, zoneId);
  }

  /**
   * Checks if a datetime matches this schedule.
   *
   * @param datetime the datetime to check
   * @return true if the datetime matches
   */
  public boolean matches(ZonedDateTime datetime) {
    ZonedDateTime dtInTz = datetime.withZoneSameInstant(zoneId);
    return Evaluator.matches(data, dtInTz, zoneId);
  }

  /**
   * Returns a lazy stream of occurrences starting after the given time.
   *
   * @param from the reference time (exclusive)
   * @return a stream of occurrences
   */
  public Stream<ZonedDateTime> occurrences(ZonedDateTime from) {
    ZonedDateTime fromInTz = from.withZoneSameInstant(zoneId);
    return Evaluator.occurrences(data, fromInTz, zoneId);
  }

  /**
   * Returns a lazy stream of occurrences where from &lt; occurrence &lt;= to.
   *
   * @param from the start time (exclusive)
   * @param to the end time (inclusive)
   * @return a stream of occurrences in the range
   */
  public Stream<ZonedDateTime> between(ZonedDateTime from, ZonedDateTime to) {
    ZonedDateTime fromInTz = from.withZoneSameInstant(zoneId);
    ZonedDateTime toInTz = to.withZoneSameInstant(zoneId);
    return Evaluator.between(data, fromInTz, toInTz, zoneId);
  }

  /**
   * Converts this schedule to a 5-field cron expression.
   *
   * @return the cron expression
   * @throws HronException if the schedule cannot be expressed as cron
   */
  public String toCron() throws HronException {
    return CronConverter.toCron(data);
  }

  /**
   * Returns the IANA timezone name, or empty if not specified.
   *
   * @return the timezone name
   */
  public Optional<String> timezone() {
    return Optional.ofNullable(data.timezone()).filter(s -> !s.isEmpty());
  }

  /**
   * Returns the canonical string representation of this schedule.
   *
   * @return the canonical form
   */
  @Override
  public String toString() {
    return Display.render(data);
  }

  /**
   * Returns the underlying schedule data.
   *
   * @return the schedule data
   */
  public ScheduleData data() {
    return data;
  }

  /** Resolve timezone, defaulting to UTC for deterministic behavior. */
  private static ZoneId resolveTimezone(String tzName) {
    if (tzName == null || tzName.isEmpty()) {
      return ZoneId.of("UTC");
    }
    return ZoneId.of(tzName);
  }
}
